[A1WAYSD] iP

A1WAYSD opened the pull request:
# DukePro
> “Beware the barrenness of a busy life.” — Socrates

Duke has now been updated to **L** 😼, a chatbot that is:

- text-based
- intelligent
- philosophical

The highlights are:

1. You can use "88" to say goodbye 👋.
2. Pop up quotations and wisdom 🧙 when chating.

Inpsired by [Momentum](https://momentumdash.com/)

Features:

- [x] Managing tasks
- [ ] Reminders
## 
Here's the `main` method:
```ruby
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

A1WAYSD committed:
Add Rename, Greet, Exit

A1WAYSD committed:
Add echo

A1WAYSD committed:
Add list

A1WAYSD committed:
Add mark feature

A1WAYSD committed:
Add task type

A1WAYSD committed:
Automate text ui testing

A1WAYSD committed:
Add error handling

A1WAYSD committed:
Add delete feature & use arraylist

A1WAYSD committed:
Add change to classes

A1WAYSD committed:
Add enums

A1WAYSD committed:
Improve code style

A1WAYSD committed:
Add save feature

A1WAYSD committed:
Merge branch 'branch-level-7'

A1WAYSD committed:
Add local date

A1WAYSD committed:
Add data.txt

A1WAYSD committed:
Merge branch 'branch-level-8'

A1WAYSD committed:
Add ui and parser

A1WAYSD committed:
Fix load/save file error

A1WAYSD committed:
Divide classes into packages

A1WAYSD committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

A1WAYSD committed:
Add Gradle for test and checkstyle

A1WAYSD committed:
Add JUnit tests

A1WAYSD committed:
Add JavaDoc

A1WAYSD committed:
Follow the Coding Standard

A1WAYSD committed:
Add find feature

A1WAYSD committed:
Merge branch 'branch-A-JavaDoc'

A1WAYSD committed:
Merge branch 'branch-A-CodingStandard'

A1WAYSD committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

A1WAYSD committed:
Improve javadoc & junit

leontan2 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    public void handleException(DukeException e) {
+        String message = e.getMessage();
+        if (message.equals("undefined")) {
+            System.out.println("⚠ Sorry! I am not able to understand you. Try another language:D");
+        } else if (message.equals("task not found")) {
+            System.out.println("⚠ Oops! Cannot find task:(");
+        } else if (message.equals("todo error")) {
+            System.out.println("⚠ Oops! Need description for the todo:(");
+        } else if (message.equals("deadline error")) {
+            System.out.println("⚠ Oops! Need description and formatted by date for the deadline:(");
+        } else if (message.equals("event error")) {
+            System.out.println("⚠ Oops! Need description, from and to date for the event:(");
+        } else {
+            System.out.println("⚠ Oops! Something went wrong:(");
+        }
```
Perhaps it would be better to use a switch statement here as there are a lot of if-else statements which may affect the clarity of the code

A1WAYSD replied:
okay will improve thanks!

leontan2 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,80 @@
+     * Loads tasks from the specified file path. Create the folder if necessary.
+     * @throws FileNotFoundException
+     */
+    public ArrayList<Task> load() throws FileNotFoundException {
```
Perhaps instead of using `load` for the name, you can use `loadFiles` for better clarity?

leontan2 commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,8 @@
+package duke.task;
+
+/**
+ * Represents the type of task.
+ */
+public enum TaskType {
+    TODO, DEADLINE, EVENT
+}
```
I like the abstraction of TaskType into a new class to represent different types of tasks instead of just storing it as a String somewhere.

leontan2 commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,34 @@
+        this.taskType = TaskType.TODO;
+    }
+
+    public ToDo(String description, boolean isDone) {
```
Perhaps for clarity, you can include some JavaDoc comments?

leontan2 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,24 @@
+        assertEquals("", new Parser(null).parse("88"));
+    }
+    @Test
+    public void parse_undefinedInput_exceptionThrown() {
```
Good adherence to the "featureUnderTest_testScenario_expectedBehavior()" as per the coding standard for test cases.

ChongWeiJie29 replied:
yep i agree with this review comment!

leontan2 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,67 @@
+        System.out.println("Generating new chat session...");
+    }
+
+    public void showSavingError() {
```
I like the descriptive naming sense for your functions, definitely helps with the clarity of what those functions do!

leontan2 reviewed (commented):
Overall code is very clean and easy to read, especially with the intuitive function names e.g. showSavingError();
Good and succinct abstraction of code into different classes which helped with the clarity.
Test cases were sufficient and handled a majority of the cases.

The issues that I had was with the convoluted if-else loop which I felt could be easily replaced with a switch statement and perhaps some missing JavaDoc comments here and there.

Anyways, LGTM!

A1WAYSD committed:
Add checkStyle & improve code

A1WAYSD committed:
Merge branch 'branch-A-CheckStyle'

Kailash201 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @param filePath the file path to load past data from.
+     */
+    public Duke(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.loadFile());
+        } catch (FileNotFoundException e) {
+            ui.showLoadingError();
+            tasks = new TaskList();
+        }
+    }
```
I like your duke method due to its simplicity which is made possible due to OOP usage.

Kailash201 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    @Override
+    public String toTxt() {
+        return this.by != null ? super.toTxt() + this.description + " | " + this.by + this.byDescription : super.toTxt()
+                + this.description + " | " + this.byDescription;
```
Is the line too long? I could not check the line length from here. If it is not, it is all good.

Kailash201 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Event extends Task {
+
+    protected String from;
+
+    protected String to;
```
extremely minor issue, do you think it will be better if there is no space between them? since the rest of your files have no space between the variables

Kailash201 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+                throw new DukeException("deadline error");
+            }
+            break;
+        default:
+            Pattern pattern = Pattern.compile("(.+) /from (.+) /to (.+)");
+            Matcher matcher = pattern.matcher(description);
+            if (matcher.matches() && !matcher.group(1).isEmpty() && !matcher.group(2).isEmpty()
+                    && !matcher.group(3).isEmpty()) {
+                this.tasks.add(new Event(matcher.group(1), matcher.group(2), matcher.group(3)));
+            } else {
+                throw new DukeException("event error");
+            }
+            break;
+        }
```
Do you think it will be better to not put event as your default case? In case there is a need to add extra features with different commands

A1WAYSD replied:
Good suggestion. Thanks!!

Kailash201 reviewed (commented):
Overall, the code was good and clean, it followed the coding standards. OOP was used well which made your code simpler to read. Just need to check your styling stays consistent throughout the code. Other than that solid work.
Keep up the good work!

A1WAYSD committed:
Follow JavaFX tutorial

ChongWeiJie29 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+public class DukeException extends Exception {
+    /**
+     * Constructs a DukeException with the specified detail message.
+     * @param message the detail message.
```
Perhaps u can consider leaving a spacing between the description and @param?

ChongWeiJie29 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,80 @@
+import duke.task.TaskList;
+import duke.task.TaskType;
+import duke.task.ToDo;
+import java.io.File;
```
Just a suggestion, but perhaps you can leave a line in between the last duke import line and first java import line to make it neater?

ChongWeiJie29 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,80 @@
+        return tasks;
+    }
+
+    private Task addTask(String description) {
```
Might be better to add a description here since the function is not trivial?

ChongWeiJie29 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+    public String addTask(TaskType taskType, String description) throws DukeException {
+        description = description.trim();
+        switch (taskType) {
+        case TODO:
```
good indentation of switch statements which is often overlooked

ChongWeiJie29 reviewed (commented):
Overall, very well organised and largely follows the coding standard. Code is also well-formatted and easy to understand through the inclusion of white spaces and line breaks. Good work

A1WAYSD committed:
Add gui

A1WAYSD committed:
Merge branch 'branch-Level-10'

A1WAYSD committed:
Use Varargs

A1WAYSD committed:
Add assertions

Various stages during program execution are unchecked.

Checking certain stages could ensure the program run correctly and
identify potential bugs. Java Asset serves as an useful tool to achieve
this.

Let's,
* check all task descriptions are valid
* ensure data storage runs successfully

Data storage is vital for the chatbot to recall history data and save
the current one. Task descriptions should not be empty.

A1WAYSD committed:
Improve code quality

Some codes need polish and refactor.

Following code standard is really important.

Let's add default to switch cases, JavaDocs and structure code logically.

A1WAYSD committed:
Merge pull request #1 from A1WAYSD/branch-A-Assertions

Add assertions

A1WAYSD committed:
Resolve conflict

A1WAYSD committed:
Merge branch 'master' into branch-A-CodeQuality

A1WAYSD committed:
Merge pull request #2 from A1WAYSD/branch-A-CodeQuality

Improve code quality

A1WAYSD committed:
Avoid magic numbers(words)

A1WAYSD committed:
Set up CI

A1WAYSD committed:
Improve code style

A1WAYSD committed:
Add reminder

A1WAYSD committed:
Merge branch 'feature-reminders'

A1WAYSD committed:
Improve GUI

A1WAYSD committed:
Add more error handling

A1WAYSD committed:
Add more automated tests

A1WAYSD committed:
Add user guide

A1WAYSD committed:
Update user guide

A1WAYSD committed:
Update JUNIT tests

A1WAYSD committed:
Fix style

A1WAYSD committed:
Fix style on ParserTest

A1WAYSD committed:
Test jar

A1WAYSD committed:
Save data in every action

A1WAYSD committed:
Update readme

A1WAYSD committed:
Print error if wrong data read

A1WAYSD committed:
Enable dialogBox expand height

A1WAYSD committed:
Update dependencies

A1WAYSD committed:
Update README.md

A1WAYSD committed:
Update README.md
